File: LeetCode/33搜索旋转排序数组.go
func search(nums []int, target int) int {
	size := len(nums)
	if size == 0 {
		return -1
	} else if size == 1 {
		if nums[0] == target {
			return 0
		}
		return -1
	}
	left, right := 0, size-1
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] == target {
			return mid
		}
		if nums[0] <= nums[mid] {
			if nums[0] <= target && target < nums[mid] {
				right = mid - 1
			} else {
				left = mid + 1
			}
		} else {
			if nums[mid] < target && target <= nums[size-1] {
				left = mid + 1
			} else {
				right = mid - 1
			}
		}
	}
	return -1
}

File: LeetCode/33搜索旋转排序数组.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int len = nums.size();
        if (len == 0) return -1;
        if (len == 1) return nums[0] == target ? 0 : -1;
        int left = 0, right = len - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) return mid;
            if (nums[0] <= nums[mid]) {
                if (nums[0] <= target && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[len - 1]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return -1;
    }
};

// 二分搜索也可以搜索部分有序，部分无序的
// 无序的再进行二分搜索，一定有部分有序，有部分无序，从而再进行二分